### ISSUE TITLE
Python 3.12.0 ModuleNotFoundError: No module named 'bottle.ext.websocket'

### ISSUE THREAD

**Issue by neelkalpa** [brentvollebregt/auto-py-to-exe#433]
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Program Files\Python312\Lib\site-packages\auto_py_to_exe\__main__.py", line 10, in <module>
    from . import ui
  File "C:\Program Files\Python312\Lib\site-packages\auto_py_to_exe\ui.py", line 6, in <module>
    import eel
  File "C:\Program Files\Python312\Lib\site-packages\eel\__init__.py", line 16, in <module>
    import bottle.ext.websocket as wbs
ModuleNotFoundError: No module named 'bottle.ext.websocket'

Tried solving like this:
pip install bottle-websocket 
pip install eel

However, it didn't work!

**Comment by brentvollebregt** [comment]
Could you try the following

1. Create a new Python virtual environment 
2. Install auto-py-to-exe in this new virtual environment using pip (make sure you are using the new virtual environment!)
3. Run the auto-py-to-exe in this virtual environment

Do you get the same issue? With you clearing the issue template out, I can't offer much more than this immediately 😕

**Comment by DLord420** [comment]
Same problem here.  Also on python 3.12.
I get the same error when trying to open the app in CMD.

**Comment by Wilsman** [comment]
just ran into the same issue. Also, not too sure how to run it in a venv 🥲

**Comment by brentvollebregt** [comment]
> It works in virtual environment. But it doesn't work outside of it.

This shows that the issue is the packages installed in your environment. 

To fix this, run one of the following:
- If you installed auto-py-to-exe using pip: `pip install auto-py-to-exe --upgrade --force-reinstall`
- If you downloaded the repo and ran auto-py-to-exe from there: `pip install -r .\requirements.txt --upgrade --force-reinstall` at the root of this project

**Comment by Wilsman** [comment]
has the issue stemmed from upgrading to Python 3.12.0?

**Comment by brentvollebregt** [comment]
> has the issue stemmed from upgrading to Python 3.12.0?

That's a good thought!

Even though @neelkalpa didn't say what Python version they were using, I think I can assume 3.12 judging by the paths. I imagine the virtual environment was the same version.

I just added Python 3.12 to the tests in this repo [and they seemed to run fine](https://github.com/brentvollebregt/auto-py-to-exe/actions/runs/6417241816) - so it looks like an upgrade to 3.12 should be ok.

**Comment by CwGmZ971** [comment]
> * pip install auto-py-to-exe --upgrade --force-reinstall

Have tried this and this has not fixed the issue, this issue is persistent with Python 3.12.0 when running auto-py-to-exe from the command prompt (CMD)

**Comment by Wilsman** [comment]
I tried the same and failed... i uploaded a pastebin of my terminal which may/may not help 🤞

[terminal - pastebin](https://pastebin.com/3z3ZavG3)

**Comment by brentvollebregt** [comment]
Thank you, I have been able to reproduce this now,

1. Installed Python 3.12
2. Setup a virtual environment: `C:\...\Python312\python.exe -m venv .venv`
3. Test version: `C:\...\.venv\Scripts\python.exe --version` gives "Python 3.12.0"
4. Install auto-py-to-exe : `C:\...\.venv\Scripts\python.exe -m pip install auto-py-to-exe`
5. Run auto-py-to-exe: `C:\...\.venv\Scripts\python.exe -m auto_py_to_exe`

I get the same as above,

```
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Brent\Downloads\t\.venv\Lib\site-packages\auto_py_to_exe\__main__.py", line 10, in <module>
    from . import ui
  File "C:\Users\Brent\Downloads\t\.venv\Lib\site-packages\auto_py_to_exe\ui.py", line 6, in <module>
    import eel
  File "C:\Users\Brent\Downloads\t\.venv\Lib\site-packages\eel\__init__.py", line 16, in <module>
    import bottle.ext.websocket as wbs
ModuleNotFoundError: No module named 'bottle.ext.websocket'
```

> I tested this with Python 3.9.9 and it seemed ok

This looks like an Eel issue, executing `C:\...\.venv\Scripts\python.exe -c "import eel"` gives the same error. I can't see any recent releases in Eel or bottle.

Interestingly, looking at how `import bottle.ext.websocket` works, it [actually imports the package](https://github.com/bottlepy/bottle/blob/master/bottle.py#L4358) `bottle-websocket`. I see I have `bottle-websocket` 0.2.9 installed so this is a bit odd that it can't import it, maybe bottle's `_ImportRedirect` is broken?

**Comment by brentvollebregt** [comment]
bottle's last release was on March 5 with version [0.12.25](https://pypi.org/project/bottle/0.12.25/). [Since then, more than 500 commits have been put on master](https://github.com/bottlepy/bottle/compare/0.12.25...master), but no release has been made to push these changes to PyPI.

I just tried the [bottle.py on master](https://github.com/bottlepy/bottle/blob/99341ff3791b2e7e705d7373e71937e9018eb081/bottle.py) in my local setup and it worked fine. So it looks like the code is available, it just hasn't been pushed to PyPI - and I can't quite see why in the repo's issues, maybe it has lost contributors with push access?

**Comment by DLord420** [comment]
Cleared PIP cache and tried `pip install auto-py-to-exe --upgrade --force-reinstall` again with no luck.

**Comment by brentvollebregt** [comment]
Yep, my bad, `pip install auto-py-to-exe --upgrade --force-reinstall` will not work. Until that bottle change comes though, it doesn't look like we can resolve this.

**Comment by Wilsman** [comment]
> bottle's last release was on March 5, 2023 with version [0.12.25](https://pypi.org/project/bottle/0.12.25/). [Since then, there have been a few changes put on master](https://github.com/bottlepy/bottle/compare/ca6762c559c5e71e0dff71dc97eb4c6b3ed9bbcd...master), but no release has been made to push these changes to PyPI.
> 
> I just tried the [bottle.py on master](https://github.com/bottlepy/bottle/blob/99341ff3791b2e7e705d7373e71937e9018eb081/bottle.py) in my local setup and it worked fine. So it looks like the code is available, it just hasn't been pushed to PyPI.

thanks for looking into this 👍

**Comment by brentvollebregt** [comment]
[The commit we need](https://github.com/bottlepy/bottle/commit/ca6762c559c5e71e0dff71dc97eb4c6b3ed9bbcd) was put on master back on June 13, 2022 but it appears that releases are not directly done master, but instead in their own release branch - so that explains how the commit was made so long ago but is not in the recent release.

**Comment by brentvollebregt** [comment]
I've created an issue at https://github.com/bottlepy/bottle/issues/1430 to let bottle know

**Comment by eLTehh** [comment]
> I've created an issue at [bottlepy/bottle#1430](https://github.com/bottlepy/bottle/issues/1430) to let bottle know

Hi, thanks for creating the issue!

For everyone else encountering the problem, quick fix I've found using the information raised by @brentvollebregt: Navigate to `Python312/Lib/site-packages/bottle.py`, and replace the code with the linked version: https://github.com/bottlepy/bottle/blob/ca6762c559c5e71e0dff71dc97eb4c6b3ed9bbcd/bottle.py

**Comment by brentvollebregt** [comment]
I have a temporary solution in #434 - this seems to work ok for me. 

Could someone else that's having this issue try this out and let me know if it works for you? If so, I'll merge it in and create a release.

**Comment by DLord420** [comment]
Tried the proposed solution and it did not work on my machine.  I just got page after page error messages ending with:

> NameError: name 'true' is not defined. Did you mean: 'True'?

**Comment by Wilsman** [comment]
> quick fix

This quick fix worked.

**Comment by brentvollebregt** [comment]
This should now be fixed in auto-py-to-exe [2.41.0](https://github.com/brentvollebregt/auto-py-to-exe/releases/tag/v2.41.0) with the shim added in #434.

I'll close this for now, but if you are still facing this issue please comment here and provide the following:

- auto-py-to-exe: `pip show auto-py-to-exe`
- Python version: `python --version`
- bottle version: `pip show bottle`

Hopefully bottle fixes this soon and we can remove the shim.
